### PR TITLE
Attempt to protect against doing things in the wrong environment

### PR DESCRIPTION
**Pull request by tijmenb**
## Context

We sometimes accidentally do stuff on production that we intended to do in development or QA. This is not good!

## Changes proposed in this pull request

This is a proposal to wrap "dangerous" features in a `protect_against_mistakes` block. This requires the user to confirm the environment, after which they will have 5 minutes to perform the dangerous tasks until they have to confirm again.

Screenshot 2021-02-17 at 13 49 07
Screenshot 2021-02-17 at 13 49 10
Screenshot 2021-02-17 at 13 49 13
Screenshot 2021-02-17 at 13 49 15

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/components/support_interface/reference_with_feedback_component.html.erb [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
+            <%= link_to support_interface_reinstate_reference_path(reference), class: 'govuk-link' do %>
+            Reinstate reference<span class="govuk-visually-hidden"> for <%= reference.name %></span>
+            <% end %>
+          </div>
+        <% elsif reference.feedback_requested? %>
+          <div class='app-summary-card__actions'>
+            <%= link_to support_interface_cancel_reference_path(reference), class: 'govuk-link' do %>
+              Cancel reference<span class="govuk-visually-hidden"> for <%= reference.name %></span>
+            <% end %>
+          </div>
```
Comment (paulrobertlloyd) [inline app/components/support_interface/reference_with_feedback_component.html.erb:16]:
```suggestion
          <div class="app-summary-card__actions">
            <%= govuk_link_to support_interface_reinstate_reference_path(reference) do %>
              Reinstate reference<span class="govuk-visually-hidden"> for <%= reference.name %></span>
            <% end %>
          </div>
        <% elsif reference.feedback_requested? %>
          <div class="app-summary-card__actions">
            <%= govuk_link_to support_interface_cancel_reference_path(reference) do %>
              Cancel reference<span class="govuk-visually-hidden"> for <%= reference.name %></span>
            <% end %>
          </div>
```

File: app/helpers/view_helper.rb [hunk 2]
```diff
@@ -146,6 +146,14 @@ def formatted_percentage(count, total)
     number_to_percentage(percentage, precision: precision)
   end
 
+  def protect_against_mistakes
+    if session[:confirmed_environment_at] && session[:confirmed_environment_at] > 10.seconds.ago #5.minutes.ago
+      yield
+    else
+      govuk_link_to '🦺 Confirm environment to make changes', support_interface_confirm_environment_path(from: request.fullpath)
```
Comment (paulrobertlloyd) [inline app/helpers/view_helper.rb:153]: Not sure we need an emoji for this… and can cause screenreader issues too if we don’t describe it properly.

File: app/views/support_interface/settings/feature_flags.html.erb [hunk 3]
```diff
@@ -27,17 +27,19 @@
...
-        <% end %>
-      <% else %>
-        <%= button_to support_interface_activate_feature_flag_path(feature_name),
-               class: 'govuk-button app-button--tertiary govuk-!-margin-bottom-0',
-               data: { confirm: "Are you sure you want to activate “#{feature_name.humanize}” in #{HostingEnvironment.environment_name.upcase}?" } do %>
-          Activate<span class="govuk-visually-hidden"> ‘<%= feature_name.humanize %>’ feature</span>
+      <% protect_against_mistakes do %>
+        <% if FeatureFlag.active?(feature_name) %>
+          <%= button_to support_interface_deactivate_feature_flag_path(feature_name),
+                 class: 'govuk-button app-button--tertiary govuk-!-margin-bottom-0',
```
Comment (paulrobertlloyd) [inline app/views/support_interface/settings/feature_flags.html.erb:33]: Think this has been fixed in a pending PR, but for review completeness:

```suggestion
          <%= govuk_button_to support_interface_deactivate_feature_flag_path(feature_name),
                 class: 'app-button--tertiary govuk-!-margin-bottom-0',
```

File: app/views/support_interface/settings/feature_flags.html.erb [hunk 4]
```diff
@@ -27,17 +27,19 @@
...
+                 data: { confirm: "Are you sure you want to deactivate “#{feature_name.humanize}” in #{HostingEnvironment.environment_name.upcase}?" } do %>
+            Deactivate<span class="govuk-visually-hidden"> ‘<%= feature_name.humanize %>’ feature</span>
+          <% end %>
+        <% else %>
+          <%= button_to support_interface_activate_feature_flag_path(feature_name),
+                 class: 'govuk-button app-button--tertiary govuk-!-margin-bottom-0',
```
Comment (paulrobertlloyd) [inline app/views/support_interface/settings/feature_flags.html.erb:39]: Ditto:

```suggestion
          <%= govuk_button_to support_interface_activate_feature_flag_path(feature_name),
                 class: 'app-button--tertiary govuk-!-margin-bottom-0',
```

**Comment by paulrobertlloyd** [comment]
❤️ this!

File: app/views/support_interface/sessions/confirm_environment.html.erb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<% content_for :title, 'Confirm environment' %>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <p class="govuk-body-l">To make sure you intend to make changes to <strong><%= HostingEnvironment.environment_name %></strong>, type <%= HostingEnvironment.environment_name %> below:</p>
+
+    <%= form_with(url: support_interface_confirm_environment_path) do |f| %>
+      <%= f.hidden_field :from, value: params[:from] %>
+      <%= f.govuk_text_field :environment, label: { text: 'Type it', size: 'm' } %>
```
Comment (paulrobertlloyd) [inline app/views/support_interface/sessions/confirm_environment.html.erb:9]: Could probably combine the guidance with the label text, maybe something like:

## Confirm action

This is the **production** environment.

Type `production` to confirm that you want to proceed
[_________________________]

Comment (paulrobertlloyd) [reply app/views/support_interface/sessions/confirm_environment.html.erb:9]: (Clearly I’m not a content designer 😂)

File: app/views/support_interface/settings/feature_flags.html.erb [hunk 6]
```diff
@@ -27,17 +27,19 @@
...
-      <% else %>
-        <%= govuk_button_to support_interface_activate_feature_flag_path(feature_name),
-               class: 'app-button--tertiary govuk-!-margin-bottom-0',
-               data: { confirm: "Are you sure you want to activate “#{feature_name.humanize}” in #{HostingEnvironment.environment_name.upcase}?" } do %>
-          Activate<span class="govuk-visually-hidden"> ‘<%= feature_name.humanize %>’ feature</span>
+      <% protect_against_mistakes do %>
+        <% if FeatureFlag.active?(feature_name) %>
+          <%= govuk_button_to support_interface_deactivate_feature_flag_path(feature_name),
+                 class: 'app-button--tertiary govuk-!-margin-bottom-0',
+                 data: { confirm: "Are you sure you want to deactivate “#{feature_name.humanize}” in #{HostingEnvironment.environment_name.upcase}?" } do %>
```
Comment (duncanjbrown) [inline app/views/support_interface/settings/feature_flags.html.erb:34]: Shall we lose the UJS data-confirm?

Comment (duncanjbrown) [reply app/views/support_interface/settings/feature_flags.html.erb:34]: (And with it UJS)

**Comment by duncanjbrown** [comment]
If I type the wrong thing into the "confirm environment" box I get stuck on the page. Should we show an error or sth?

**Comment by tijmenb** [comment]
@duncanjbrown updated PR!

File: app/forms/support_interface/confirm_environment.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+module SupportInterface
+  class ConfirmEnvironment
+    include ActiveModel::Model
+    attr_accessor :from, :environment
+
+    validate :correct_environment
+
+    def correct_environment
+      if environment != HostingEnvironment.environment_name
+        errors[:environment] << "That’s not ’#{HostingEnvironment.environment_name}’!"
```
Comment (duncanjbrown) [inline app/forms/support_interface/confirm_environment.rb:10]: 👌

**Review by duncanjbrown — APPROVED** [review]
🎉

**Pull request merged**